volume/store/store.go

```diff
@@ -14,175 +16,290 @@ var (
+	// ErrNameConflict is a typed error returned on create when a volume exists with the given name, but for a different driver
+	ErrNameConflict = errors.New("conflict: volume name must be unique")
 )
 
 // New initializes a VolumeStore to keep
 // reference counting of volumes in the system.
 func New() *VolumeStore {
 	return &VolumeStore{
-		vols: make(map[string]*volumeCounter),
+		names:      make(map[string]string),
+		refs:       make(map[string][]string),
+		nameLocker: locker.New(),
 	}
 }
 
 // VolumeStore is a struct that stores the list of volumes available and keeps track of their usage counts
 type VolumeStore struct {
-	vols map[string]*volumeCounter
-	mu   sync.Mutex
+	names      map[string]string
+	refs       map[string][]string
+	nameLocker *locker.Locker
+	mu         sync.Mutex
 }
 
-// volumeCounter keeps track of references to a volume
-type volumeCounter struct {
-	volume.Volume
-	count uint
-}
+// List proxies to all registered volume drivers to get the full list of volumes
+func (s *VolumeStore) List() ([]volume.Volume, []string, error) {
+	vols, warnings, err := s.list()
+	if err != nil {
+		return nil, nil, err
+	}
+	var out []volume.Volume
 
-// AddAll adds a list of volumes to the store
-func (s *VolumeStore) AddAll(vols []volume.Volume) {
+	s.mu.Lock()
 	for _, v := range vols {
-		s.vols[v.Name()] = &volumeCounter{v, 0}
+		s.nameLocker.Lock(v.Name())
+		driverName, exists := s.names[v.Name()]
+		if !exists {
+			s.names[v.Name()] = v.DriverName()
+		}
+		if exists && driverName != v.DriverName() {
+			logrus.Warnf("Volume name %s already exists for driver %s, not including volume returned by %s", v.Name(), driverName, v.DriverName())
+			s.nameLocker.Unlock(v.Name())
+			continue
+		}
+
+		out = append(out, v)
+		s.nameLocker.Unlock(v.Name())
+	}
+	s.mu.Unlock()
+	return vols, warnings, nil
+}
+
+func (s *VolumeStore) list() ([]volume.Volume, []string, error) {
+	drivers, err := volumedrivers.GetAllDrivers()
+	if err != nil {
+		return nil, nil, err
+	}
+	var (
+		ls       []volume.Volume
+		warnings []string
+	)
+
+	type vols struct {
+		vols []volume.Volume
+		err  error
+	}
+	chVols := make(chan vols, len(drivers))
+
+	for _, vd := range drivers {
+		logrus.Debugf("probing %s", vd.Name())
+		go func(d volume.Driver) {
+			vs, err := d.List()
+			if err != nil {
+				chVols <- vols{err: fmt.Errorf("Error getting volume list for driver %s: %v", d.Name(), err)}
+			}
+			chVols <- vols{vols: vs}
+		}(vd)
+	}
+
+	for i := 0; i < len(drivers); i++ {
+		vs := <-chVols
+
+		if vs.err != nil {
+			warnings = append(warnings, vs.err.Error())
+			logrus.Warn(vs.err)
+			continue
+		}
+		ls = append(ls, vs.vols...)
+	}
+	return ls, warnings, nil
+}
+
+// CreateWithRef creates a volume with the given name and driver and stores the ref
+func (s *VolumeStore) CreateWithRef(name, driverName, ref string, opts map[string]string) (volume.Volume, error) {
+	s.nameLocker.Lock(name)
+	defer s.nameLocker.Unlock(name)
```

Comment:
As is currently designed, it's important for the name lock to be locked while modifying anything so that nothing is trying to read/do stuff at the same time.
We could use the global lock in all cases, but then we have the same issue we're trying to solve with the nested locks.